openmeter/billing/service/lineservice/linebase.go

```diff
@@ -0,0 +1,229 @@
+
+func (l lineBase) ToEntity() billingentity.Line {
+	return l.line
+}
+
+func (l lineBase) ID() string {
+	return l.line.ID
+}
+
+func (l lineBase) InvoiceID() string {
+	return l.line.InvoiceID
+}
+
+func (l lineBase) Currency() currencyx.Code {
+	return l.line.Currency
+}
+
+func (l lineBase) Period() billingentity.Period {
+	return l.line.Period
+}
+
+func (l lineBase) Status() billingentity.InvoiceLineStatus {
+	return l.line.Status
+}
+
+func (l lineBase) HasParent() bool {
+	return l.line.ParentLineID != nil
+}
+
+func (l lineBase) Validate(ctx context.Context, invoice *billingentity.Invoice) error {
+	if l.line.Currency != invoice.Currency || l.line.Currency == "" {
+		return billingentity.ValidationError{
+			Err: billingentity.ErrInvoiceLineCurrencyMismatch,
+		}
+	}
+
+	return nil
+}
+
+func (l lineBase) Save(ctx context.Context) (Line, error) {
+	line, err := l.service.BillingAdapter.UpdateInvoiceLine(ctx, billing.UpdateInvoiceLineAdapterInput(l.line))
+	if err != nil {
+		return nil, fmt.Errorf("updating invoice line: %w", err)
+	}
+
+	return l.service.FromEntity(line)
+}
+
+func (l lineBase) Service() *Service {
+	return l.service
+}
+
+func (l lineBase) CloneForCreate(in UpdateInput) Line {
+	l.line.ID = ""
+	l.line.CreatedAt = time.Time{}
+	l.line.UpdatedAt = time.Time{}
+
+	return l.update(in)
+}
+
+func (l lineBase) Update(in UpdateInput) Line {
+	return l.update(in)
+}
+
+func (l lineBase) update(in UpdateInput) Line {
+	in.apply(&l.line)
+
+	// Let's update the parent line
+	if in.ParentLine != nil {
+		newParentLine := *in.ParentLine
+		if newParentLine == nil {
+			l.line.ParentLineID = nil
+			l.line.ParentLine = nil
+		} else {
+			l.line.ParentLineID = lo.ToPtr(newParentLine.ID())
+			l.line.ParentLine = lo.ToPtr(newParentLine.ToEntity())
+		}
+	}
+
+	// Let's ignore the error here as we don't allow for any type updates
+	svc, _ := l.service.FromEntity(l.line)
+
+	return svc
+}
+
+func (l lineBase) Split(ctx context.Context, splitAt time.Time) (SplitResult, error) {
+	// We only split valid lines; split etc. lines are not supported
+	if l.line.Status != billingentity.InvoiceLineStatusValid {
+		return SplitResult{}, fmt.Errorf("line[%s]: line is not valid", l.line.ID)
+	}
+
+	if !l.line.Period.Contains(splitAt) {
+		return SplitResult{}, fmt.Errorf("line[%s]: splitAt is not within the line period", l.line.ID)
+	}
+
+	if !l.HasParent() {
+		parentLine, err := l.Update(UpdateInput{
+			Status: billingentity.InvoiceLineStatusSplit,
+		}).Save(ctx)
+		if err != nil {
+			return SplitResult{}, fmt.Errorf("saving parent line: %w", err)
+		}
+
+		// Let's create the child lines
+		preSplitAtLine := l.CloneForCreate(UpdateInput{
+			ParentLine: &parentLine,
+			Status:     billingentity.InvoiceLineStatusValid,
+			PeriodEnd:  splitAt,
+			InvoiceAt:  splitAt,
+		})
+
+		postSplitAtLine := l.CloneForCreate(UpdateInput{
+			ParentLine:  &parentLine,
+			Status:      billingentity.InvoiceLineStatusValid,
+			PeriodStart: splitAt,
```

Comment:
just double checking: is this is also splitAt and and previous ends at splitAt one should be exclusive when we query usage

Reply:
The plan was that:

```
// Period represents a time period, in billing the time period is always interpreted as
// [start, end) (i.e. start is inclusive, end is exclusive).
```